src/eval/mod.rs

```diff
@@ -1184,6 +1185,102 @@ impl Eval for ast::Closure {
     }
 }
 
+impl ast::Pattern {
+    // Destruct the given value into the pattern.
+    pub fn define(&self, vm: &mut Vm, value: Value) -> SourceResult<Value> {
+        match self.kind() {
+            ast::PatternKind::Ident(ident) => {
+                vm.define(ident, value);
+                Ok(Value::None)
+            }
+            ast::PatternKind::Destructure(pattern) => {
+                match value {
+                    Value::Array(value) => {
+                        let mut i = 0;
+                        for p in &pattern {
+                            match p {
+                                ast::DestructuringKind::Ident(ident) => {
+                                    let Ok(v) = value.at(i) else {
+                                        bail!(ident.span(), "not enough elements to destructure");
+                                    };
+                                    vm.define(ident.clone(), v.clone());
+                                    i += 1;
+                                }
+                                ast::DestructuringKind::Sink(ident) => {
+                                    let sink_size =
+                                        value.len() as i64 - pattern.len() as i64 + 1;
+                                    if sink_size < 0 {
+                                        bail!(
+                                            self.span(),
+                                            "not enough elements to destructure"
+                                        );
+                                    }
+                                    let Ok(v) = value.slice(i, Some(i + sink_size)) else {
+                                        bail!(self.span(), "not enough elements to destructure");
+                                    };
+                                    if let Some(ident) = ident {
+                                        vm.define(ident.clone(), v.clone());
+                                    }
+                                    i += sink_size;
```

Comment:
by doing something like `value.len().checked_sub(pattern.len()).and_then(|c| value.slice(i, Some(i + c as i64)))` you can get rid of the duplicated error message